go/store/prolly/artifact_map.go

```diff
@@ -0,0 +1,409 @@
+		if err == io.EOF {
+			break
+		}
+
+		dErr := edt.Delete(ctx, art.ArtKey)
+		if dErr != nil {
+			return ArtifactMap{}, dErr
+		}
+	}
+
+	return edt.Flush(ctx)
+}
+
+// CountOfType returns the number of artifacts of |artType|.
+func (m ArtifactMap) CountOfType(ctx context.Context, artType ArtifactType) (cnt uint64, err error) {
+	itr, err := m.iterAllOfType(ctx, artType)
+	if err != nil {
+		return 0, err
+	}
+	for err != io.EOF {
+		_, err = itr.Next(ctx)
+		if err != nil && err != io.EOF {
+			return 0, err
+		}
+		cnt++
+	}
+	return cnt, nil
+}
+
+func (m ArtifactMap) iterAllOfType(ctx context.Context, artType ArtifactType) (artifactTypeIter, error) {
+	itr, err := m.IterAll(ctx)
+	if err != nil {
+		return artifactTypeIter{}, err
+	}
+	return artifactTypeIter{itr, artType}, nil
+}
+
+func MergeArtifactMaps(ctx context.Context, left, right, base ArtifactMap, cb tree.CollisionFn) (ArtifactMap, error) {
+	serializer := message.ProllyMapSerializer{Pool: left.tuples.ns.Pool()}
+	tuples, err := mergeOrderedTrees(ctx, left.tuples, right.tuples, base.tuples, cb, serializer)
+	if err != nil {
+		return ArtifactMap{}, err
+	}
+
+	return ArtifactMap{
+		tuples:  tuples,
+		keyDesc: base.keyDesc,
+		valDesc: base.valDesc,
+	}, nil
+}
+
+type ArtifactsEditor struct {
+	mut        MutableMap
+	srcKeyDesc val.TupleDesc
+}
+
+func (wr ArtifactsEditor) Add(ctx context.Context, key val.Tuple, val val.Tuple) error {
+	return wr.mut.Put(ctx, key, val)
+}
+
+func (wr ArtifactsEditor) Delete(ctx context.Context, key val.Tuple) error {
+	return wr.mut.Delete(ctx, key)
+}
+
+func (wr ArtifactsEditor) Flush(ctx context.Context) (ArtifactMap, error) {
+	m, err := wr.mut.Map(ctx)
+	if err != nil {
+		return ArtifactMap{}, err
+	}
+
+	return ArtifactMap{
+		tuples:     m.tuples,
+		srcKeyDesc: wr.srcKeyDesc,
+		keyDesc:    wr.mut.keyDesc,
+		valDesc:    wr.mut.valDesc,
+	}, nil
+}
+
+// ConflictArtifactIter iters all the conflicts in ArtifactMap.
+type ConflictArtifactIter struct {
+	itr artifactTypeIter
+}
+
+func (itr *ConflictArtifactIter) Next(ctx context.Context) (ConflictArtifact, error) {
+	art, err := itr.itr.Next(ctx)
+	if err != nil {
+		return ConflictArtifact{}, err
+	}
+
+	var parsedMeta ConflictMetadata
+	err = json.Unmarshal(art.Metadata, &parsedMeta)
+	if err != nil {
+		return ConflictArtifact{}, err
+	}
+
+	return ConflictArtifact{
+		Key:        art.Key,
+		HeadCmHash: art.HeadCmHash,
+		Metadata:   parsedMeta,
+	}, nil
+}
+
+// ConflictArtifact is the decoded conflict from the artifacts table
+type ConflictArtifact struct {
+	Key        val.Tuple
+	HeadCmHash []byte
+	Metadata   ConflictMetadata
+}
+
+// ConflictMetadata is the json metadata associated with a conflict
+type ConflictMetadata struct {
+	// BaseTblHash is the target hash of the table holding the base value for the conflict
+	BaseTblHash []byte `json:"bc"`
+	// TheirTblHash is the target hash of the table holding the their value for the conflict
+	TheirTblHash []byte `json:"tc"`
+}
```

Comment:
I don't see a need to store the data as json specifically. The only thing we do need is flexibility between the different artifact types.

To achieve that, I guess the encoding for the artifact value has to be a ByteStringEnc instead of a JSONEnc?